src/extension/coqPilot.ts

```diff
@@ -81,70 +75,126 @@ export class CoqPilot {
             this.performCompletionForAllAdmits.bind(this)
         );
 
+        const toggleCommand = `toggle_current_session`;
+        this.registerEditorCommand(
+            toggleCommand,
+            this.toggleCurrentSession.bind(this)
+        );
+        this.healthStatusIndicator = new HealthStatusIndicator(
+            `${pluginId}.${toggleCommand}`,
+            vscodeExtensionContext
+        );
+        this.healthStatusIndicator.updateStatusBar(true);
+
         this.vscodeExtensionContext.subscriptions.push(this);
     }
 
+    static async create(vscodeExtensionContext: ExtensionContext) {
+        const globalExtensionState = new GlobalExtensionState();
+        const sessionExtensionState = await SessionExtensionState.create(
+            globalExtensionState.logOutputChannel,
+            globalExtensionState.eventLogger
+        );
+        return new CoqPilot(
+            vscodeExtensionContext,
+            globalExtensionState,
+            sessionExtensionState
+        );
+    }
+
     async performCompletionUnderCursor(editor: TextEditor) {
         const cursorPosition = editor.selection.active;
         this.performSpecificCompletionsWithProgress(
             (hole) => positionInRange(cursorPosition, hole.range),
-            editor
+            editor,
+            this.sessionExtensionState.abortController.signal
         );
     }
 
     async performCompletionInSelection(editor: TextEditor) {
         const selection = editor.selection;
         this.performSpecificCompletionsWithProgress(
             (hole) => selection.contains(toVSCodePosition(hole.range.start)),
-            editor
+            editor,
+            this.sessionExtensionState.abortController.signal
         );
     }
 
     async performCompletionForAllAdmits(editor: TextEditor) {
-        this.performSpecificCompletionsWithProgress((_hole) => true, editor);
+        this.performSpecificCompletionsWithProgress(
+            (_hole) => true,
+            editor,
+            this.sessionExtensionState.abortController.signal
+        );
+    }
+
+    async toggleCurrentSession() {
+        if (this.globalExtensionState.hasActiveSession) {
```

Comment:
I'm not sure if this `if` is enough: what will happen, if the user just _spams_ the `toggle` command?

Reply:
I don't want such users.............................

Reply:
What if user drops a nuke on his PC?

Reply:
You mean what happens if the client is still in the process of being created and the user switches the button on/off? Client boots quite fast, I don't see any sense in creating some new state like _bootingUp_